packages/wallet/ui/src/components/WalletConnection.jsx

```diff
@@ -71,76 +62,70 @@ const getAccessToken = () => {
-  const [backendCancel, setBackendCancel] = useState(null);
-  const onWalletState = useCallback(
-    ev => {
-      const { walletConnection, state } = ev.detail;
-      setConnectionState(state);
-      switch (state) {
-        case 'idle': {
-          if (backendCancel) {
-            backendCancel();
-          }
-
-          // This is one of the only methods that the wallet connection facet allows.
-          // It connects asynchronously, but you can use promise pipelining immediately.
-          /** @type {ERef<WalletBridge>} */
-          const bridge = E(walletConnection).getAdminBootstrap(
-            getAccessToken(),
-          );
-          const { backendIt, cancel } = makeBackendFromWalletBridge(bridge);
-
-          // Need to thunk the cancel function, or it will be called immediately.
-          setBackendCancel(() => cancel);
-
-          observeIterator(backendIt, {
-            updateState: setBackend,
-            fail: () => {
-              setBackend(null);
-            },
-          });
-          break;
-        }
-        case 'error': {
-          console.log('error', ev.detail);
-          // In case of an error, reset to 'idle'.
-          // Backoff or other retry strategies would go here instead of immediate reset.
-          E(walletConnection).reset();
-          break;
-        }
-        default:
+  const [wc, setWC] = useState(null);
+
+  let cancelled = null;
+  const onWalletState = ev => {
+    if (cancelled) {
+      return;
+    }
+    const { walletConnection: newWC, state } = ev.detail;
+    setConnectionState(state);
+    if (!wc) {
+      setWC(newWC);
+    }
+  };
+
+  useEffect(() => {
+    if (!wc) {
+      return () => {};
+    }
+    const bridge = E(wc).getAdminBootstrap(
+      getAccessToken(),
+      makeFixedWebSocketConnector(walletConnection.url),
+    );
+
+    const { backendIt, cancel } = makeBackendFromWalletBridge(bridge);
+    const rethrowIfNotCancelled = e => {
+      if (!cancelled) {
+        throw e;
       }
-    },
-    [backendCancel],
-  );
+    };
+
+    observeIterator(backendIt, {
+      updateState: be => {
+        if (cancelled) {
+          throw Error('cancelled');
+        }
+        setBackend(be);
+      },
+    }).catch(rethrowIfNotCancelled);
+
+    return () => {
+      cancelled = true;
+      disconnect();
+      cancel();
+    };
+  }, [wc]);
 
   return (
-    <div className={clsx('Connector', classes.connector)}>
-      <Tooltip
-        title={connectionState === 'bridged' ? 'Connected' : 'Disconnected'}
-      >
-        <IconButton
-          size="medium"
-          color={connectionState === 'bridged' ? 'primary' : 'secondary'}
-        >
-          <Public fontSize="inherit">{connectionState}</Public>
-        </IconButton>
-      </Tooltip>
-      <AgoricWalletConnection
-        onState={onWalletState}
-        className={classes.hidden}
-      />
-    </div>
+    <AgoricWalletConnection
+      onState={onWalletState}
```

Comment:
not set in this PR but `onState` is a confusing property name since `state` has a quite specific meaning in React